torch/onnx/_internal/fx/passes/type_promotion.py

```diff
@@ -920,3 +947,378 @@ def get_rule(
     ) -> Optional[TypePromotionRule]:
         """Get type promotion rule for a python op under 'torch.ops.<namespace>'."""
         return self._rule_table.get(str(py_op), None)
+
+
+@_beartype.beartype
+def get_type_promotion_rule(
+    diagnostic: diagnostics.Diagnostic,
+    node: torch.fx.Node,
+    type_promotion_table: TypePromotionTable,
+) -> Optional[TypePromotionRule]:
+    """Get type promotion rule for a node.
+
+    Args:
+        diagnostic: Diagnostic object.
+        node: Node to get type promotion rule for.
+        type_promotion_table: Type promotion table.
+
+    Returns:
+        Type promotion rule for the node.
+    """
+    op = node.target
+    if not isinstance(op, torch._ops.OpOverload):
+        # TODO(bowbao): diagnostic.emit api.
+        diagnostic.with_additional_message(
+            f"Skipped. node.target is not OpOverload. Got type: {type(op)}"
+        )
+        return None
+    if (rule := type_promotion_table.get_rule(op.overloadpacket)) is None:
+        diagnostic.with_additional_message(
+            f"Skipped. Cannot find type promotion rule for op: {op}"
+        )
+        return None
+
+    diagnostic.with_additional_message(f"Found type promotion rule: {rule}")
+    return rule
+
+
+class _OpTraceDispatchMode(_python_dispatch.TorchDispatchMode):
+    """Trace ops that were dispatched."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.traced_ops = []
+
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        self.traced_ops.append(func)
+        return func(*args, **kwargs)
+
+
+@_beartype.beartype
+def find_compatible_op_overload(
```

Comment:
So this is like a tiny dispatcher?

Reply:
Indeed, confusing. Any chance we could reuse or extend the current our dispatcher for this?

Reply:
I'll expand the docstring. This is indeed a tricky problem. I have tried to solve in multiple different ways. 
The current one is the most elegant and simple solution.

Reply:
Looks like when we type promote(change) args/kwargs, the opoverloadpacket needs to be dispatched again in FX graph.

I wonder if they have a schema matching system that we could use.

Reply:
> I wonder if they have a schema matching system that we could use.

afaik there's only a lower level api that checks if two schemas are compatible to each other, and it is painful to use in this scenario since we need to create a new schema from args/kwargs and check existing schemas for matches. And furthermore due to the ordering of check the schema we found might disagree with torch dispatcher. I found it easier to let `__torch_dispatch__` tell us which opoverload it dispatched to in the end.